src/axom/inlet/Table.hpp

```diff
@@ -19,17 +19,184 @@
 #include <functional>
 #include <unordered_map>
 #include <tuple>
+#include <type_traits>
+
+#include "fmt/fmt.hpp"
 
 #include "axom/inlet/Field.hpp"
 #include "axom/inlet/Reader.hpp"
 #include "axom/inlet/SchemaCreator.hpp"
 
 #include "axom/sidre.hpp"
 
+/*!
+ *******************************************************************************
+ * \brief Prototype for user-defined types wishing to define a by-value read
+ * from inlet with axom::inlet::Table::get(const std::string&)
+ * 
+ * This is the only way of reading in a non-default-constructible type from
+ * Inlet.  Usage of axom::inlet::Table::get(const std::string&) without
+ * specializing this function will result in a linker error.
+ * 
+ * \see axom::inlet::Table::get(const std::string&)
+ * 
+ * \param [in] base The Inlet table representing the root of the object
+ * in which all object fields are children
+ *******************************************************************************
+ */
+template <typename T>
+T from_inlet(axom::inlet::Table& base);
+
 namespace axom
 {
 namespace inlet
 {
+/*!
+ *******************************************************************************
+ * \class is_lua_primitive
+ *
+ * \brief A type trait for checking if a type is isomorphic to a Lua primitive
+ * \tparam T The type to check
+ *******************************************************************************
+ */
+template <typename T>
+struct is_lua_primitive
+{
+  using BaseType = typename std::decay<T>::type;
+  static constexpr bool value = std::is_same<BaseType, bool>::value ||
+    std::is_same<BaseType, int>::value || std::is_same<BaseType, double>::value ||
+    std::is_same<BaseType, std::string>::value;
+};
+
+/*!
+ *******************************************************************************
+ * \class is_lua_primitive
+ *
+ * \brief A type trait for checking if a type is isomorphic to an array of Lua
+ * primitives
+ * \tparam T The type to check
+ *******************************************************************************
+ */
+template <typename T>
+struct is_lua_primitive_array
+{
+  static constexpr bool value = false;
+};
+
+// If it's an unordered map whose value type is a lua primitive,
+// assume that it's an array
+template <typename T>
+struct is_lua_primitive_array<std::unordered_map<int, T>>
+{
+  static constexpr bool value = is_lua_primitive<T>::value;
+};
+
+/*!
+ *******************************************************************************
+ * \class Proxy
+ *
+ * \brief Provides a uniform interface for access and conversion to primitive
+ * and user-defined types
+ *
+ * \see Inlet Field
+ * \see Inlet Table
+ *******************************************************************************
+ */
+class Proxy
```

Comment:
See earlier comment about whether `Proxy` is a good user-facing name for this type.